staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/oidc/oidc.go

```diff
@@ -268,6 +273,28 @@ func New(opts Options) (*Authenticator, error) {
 		client = &http.Client{Transport: tr, Timeout: 30 * time.Second}
 	}
 
+	// If the discovery URL is set in authentication configuration, we set up a
+	// roundTripper to rewrite the issuer URL to the discovery URL for the
+	// /.well-known/openid-configuration endpoint. This is useful for self-hosted
+	// providers, for example, providers that run on top of Kubernetes itself.
+	if len(opts.JWTAuthenticator.Issuer.DiscoveryURL) > 0 {
+		discoveryURL, err := url.Parse(opts.JWTAuthenticator.Issuer.DiscoveryURL)
+		if err != nil {
+			return nil, fmt.Errorf("oidc: invalid discovery URL: %w", err)
+		}
+
+		clientWithDiscoveryURL := *client
+		baseTransport := clientWithDiscoveryURL.Transport
+		if baseTransport == nil {
+			baseTransport = http.DefaultTransport
+		}
+		// This matches the url construction in oidc.NewProvider as of go-oidc v2.2.1.
+		// xref: https://github.com/coreos/go-oidc/blob/40cd342c4a2076195294612a834d11df23c1b25a/oidc.go#L114
+		urlToRewrite := strings.TrimSuffix(opts.JWTAuthenticator.Issuer.URL, "/") + wellKnownEndpointPath
```

Comment:
can we construct the full target discovery URL at startup:

```go
discoveryURL.Path = path.Join(discoveryURL.Path, wellKnownEndpointPath)
```

Reply:
This results in
```bash
&url.URL{Scheme:"https", Opaque:"", User:(*url.Userinfo)(nil), Host:"127.0.0.1:52645", Path:".well-known/openid-configuration", RawPath:"", OmitHost:false, ForceQuery:false, RawQuery:"", Fragment:"", RawFragment:""}
```

with the path missing the prefix slash.

Reply:
I think that would actually work when it got stringified back out, but if you really want the leading slash in the internal representation, you can do this:

```go
discoveryURL.Path = path.Join(discoveryURL.Path, wellKnownEndpointPath)
```

Reply:
updated the PR with this